### PR TITLE
docs: stream guide

### PR DESCRIPTION
**Pull request by hairyhum**
Guide to use streams protocol and client usage.

### CODE REVIEW

**Comment by mrinalwadhwa** [comment]
let's put this in this path `documentation/guides/drafts` so we can merge them to develop early before they are fully ready to be part of the get-started flow.

**Comment by hairyhum** [comment]
@antoinevg There are elixir examples which I used in the streams community call demo here: https://github.com/ockam-network/ockam/tree/develop/implementations/elixir/ockam/ockam/lib/ockam/examples/stream
I showed it in 3 stages: stream APIs, stream client and bi-directional communication.
I think for the guide we can start with stream client skipping stream APIs. Especially since the stream client in rust currently  has two layers: 
- stream/consumer.rs stream/publisher.rs 
- stream/mod.rs

We can shape the guide around these two layers.

**Comment by hairyhum** [comment]
We also already have bi-directional stream examples in rust here: https://github.com/ockam-network/ockam/tree/14b05ad86278c63a76a3f76d7f645daadbdca953/examples/rust/get_started/examples

**Comment by hairyhum** [comment]
@jared-s @mrinalwadhwa would appreciate feedback on the guides content.

File: documentation/guides/drafts/rust/get-started/11-streams/README.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+## Stream client
+
+To hide complexity of stream protocol interaction, Ockam provides a Stream Client API:
+
+
+**TODO: check this API**
+
+```
+Stream::connect(route, ourgoing_stream, incoming_stream)
```
Comment (antoinevg) [inline documentation/guides/drafts/rust/get-started/11-streams/README.md:31]: `s/ourgoing_stream/outgoing_stream`

File: documentation/guides/drafts/rust/get-started/11-streams/README.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+title: Reliable message delivery with Streams
+```
+
+# Reliable message delivery with Streams
+
+## Introduction
+
+In the previous guides we were sending messages to remote workers without much delivery guarantees.
+The workers were sending messages hoping that receiving end would receive them. This messaging mode is
+sometimes called "at most once delivery"
```
Comment (jared-s) [inline documentation/guides/drafts/rust/get-started/11-streams/README.md:11]:
```suggestion
sometimes called "at most once delivery".
```

File: documentation/guides/drafts/rust/get-started/11-streams/README.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+In real life distributed systems are constantly experiencing network interruptions while workers and devices
+themselves may crash and restart.
+
+To address that message brokers introduce message buffers or logs, which store messages and can re-deliver
+them after failures.
+Ockam Hub integrates with message brokers using Ockam Streams.
+To communicate with Ockam Streams on the Hub, the application can use the Ockam Streams Protocol.
```
Comment (jared-s) [inline documentation/guides/drafts/rust/get-started/11-streams/README.md:19]:
```suggestion
To address that message brokers introduce message buffers or logs, which store messages and can re-deliver them after failures. Ockam Hub integrates with message brokers using Ockam Streams. To communicate with Ockam Streams on the Hub, the application can use the Ockam Streams Protocol.
```

File: documentation/guides/drafts/rust/get-started/11-streams/README.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    ctx.start_worker("echoer", Echoer).await?;
+
+    // Create the stream
+    Stream::new(&ctx)?
+        .stream_service("stream")
+        .index_service("stream_index")
+        .client_id("stream-over-cloud-node-initiator")
+        .with_interval(Duration::from_millis(100))
+        .connect(
+            Route::new().append_t(TCP, "127.0.0.1:4000"),
```
Comment (jared-s) [inline documentation/guides/drafts/rust/get-started/11-streams/README.md:79]:
```suggestion
            route![(TCP, "127.0.0.1:4000")],
```
Needs use in previous suggestion

File: documentation/guides/drafts/rust/get-started/11-streams/README.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+**TODO: do we want to call it over-local-node?**
+
+```
+touch examples/13-stream-over-cloud-node-responder.rs
+```
+
+Add the following code to this file:
+
+```rust
+use ockam::{stream::Stream, Context, Result, Route, TcpTransport, TCP};
```
Comment (jared-s) [inline documentation/guides/drafts/rust/get-started/11-streams/README.md:60]:
```suggestion
use ockam::{stream::Stream, Context, Result, route, TcpTransport, TCP};
```
For importing the macro suggestion below

File: documentation/guides/drafts/rust/get-started/11-streams/README.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    let tcp = TcpTransport::create(&ctx).await?;
+    tcp.connect("127.0.0.1:4000").await?;
+
+    let (tx, _rx) = Stream::new(&ctx)?
+        .stream_service("stream")
+        .index_service("stream_index")
+        .client_id("stream-over-cloud-node-initiator")
+        .with_interval(Duration::from_millis(100))
+        .connect(
+            Route::new().append_t(TCP, "127.0.0.1:4000"),
```
Comment (jared-s) [inline documentation/guides/drafts/rust/get-started/11-streams/README.md:119]:
```suggestion
            route![(TCP, "127.0.0.1:4000")],
```

File: documentation/guides/drafts/rust/get-started/12-secure-channel-over-streams/README.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+In this guide, we're going to combine that and send messages over streams using secure channels.
+This way we'll achieve both end-to-end encryption and reliable message delivery.
+
+
+## App worker
+
+We're going to run two nodes communicating via Ockam Hub. 
+One of them will run the secure channel listener and the echoer worker,
+another will run the secure channel initiator and send the message to echoer.
```
Comment (jared-s) [inline documentation/guides/drafts/rust/get-started/12-secure-channel-over-streams/README.md:20]:
```suggestion
We're going to run two nodes communicating via Ockam Hub. One of them will run the secure channel listener and the echoer worker, another will run the secure channel initiator and send the message to echoer.
```

File: documentation/guides/drafts/rust/get-started/12-secure-channel-over-streams/README.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    SecureChannel::create_listener(&ctx, "secure_channel_listener", &vault).await?;
+
+    // Create a bi-directional stream
+    Stream::new(&ctx)?
+        .stream_service("stream")
+        .index_service("stream_index")
+        .client_id("secure-channel-over-stream-over-cloud-node-responder")
+        .with_interval(Duration::from_millis(100))
+        .connect(
+            Route::new().append_t(TCP, "127.0.0.1:4000"),
```
Comment (jared-s) [inline documentation/guides/drafts/rust/get-started/12-secure-channel-over-streams/README.md:56]:
```suggestion
            route![(TCP, "127.0.0.1:4000")],
```

File: documentation/guides/drafts/rust/get-started/12-secure-channel-over-streams/README.md [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    let vault = Vault::create(&ctx)?;
+
+    // Create a bi-directional stream
+    let (tx, _) = Stream::new(&ctx)?
+        .stream_service("stream")
+        .index_service("stream_index")
+        .client_id("secure-channel-over-stream-over-cloud-node-initiator")
+        .with_interval(Duration::from_millis(100))
+        .connect(
+            Route::new().append_t(TCP, "127.0.0.1:4000"),
```
Comment (jared-s) [inline documentation/guides/drafts/rust/get-started/12-secure-channel-over-streams/README.md:101]:
```suggestion
            route![(TCP, "127.0.0.1:4000")],
```

**Review by jared-s — APPROVED** [review]
Content is great! A few minor technical suggestions. Some lines of copy could maybe be combined into paragraphs to allow more flexible word and line breaking.

**Comment by hairyhum** [comment]
This branch is based on `guides_draft`.

Commits will need to be squashed before merge.

**Comment by mrinalwadhwa** [comment]
@hairyhum could you cleanup this branch so it only has stream related guide(s) and base it on develop.
I will be cherry picking content from guides_draft and landing it into develop ... so it would be great that PR can land independently of guides_draft branch

File: examples/rust/ockam_kafka/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+RUN apt-get -y install apt-utils
+RUN apt-get -y upgrade
+RUN apt-get -y install clang lld curl git pkg-config libssl-dev
+RUN curl https://sh.rustup.rs -sSf | sh -s -- -y
+
+COPY . ./
+
+RUN . ~/.cargo/env && cargo build
+
+CMD . ~/.cargo/env && cargo run --example example
```
Comment (hairyhum) [inline examples/rust/ockam_kafka/Dockerfile:13]: Using `cargo run --example` here.
We can also build a binary.

File: examples/rust/ockam_kafka/examples/example.rs [hunk 11]
```diff
@@ -0,0 +1,28 @@
+// This node routes a message, to a worker on a cloud node, over the tcp transport.
+
+use ockam::{route, Context, Result, TcpTransport, TCP};
```
Comment (hairyhum) [inline examples/rust/ockam_kafka/examples/example.rs:3]: Using the Hub example for demonstration.
This file needs to contain the secure channel via streams example

**Comment by hairyhum** [comment]
@antoinevg created a base project for kafka example and a dockerfile to build it. Please put the example code there.

**Comment by hairyhum** [comment]
Squashed the commits.

**Review by mrinalwadhwa — APPROVED** [review]

**Pull request merged**